### PR TITLE
fix handle quotes between links, bold text and other text elements

### LINKED ISSUE #74: curly quotes go the wrong way before a link
If I run the following

```js
import { remark } from 'remark'
import smartypants from 'remark-smartypants'

const result = await remark()
  .use(smartypants)
  .process('"wow". go to "[google](https://www.google.com/)" today.')

console.log(String(result))
```

the output is:

```
“wow”. go to ”[google](https://www.google.com/)” today.
```

Note that the curly quote before the `google` link goes the wrong way – it's a closing quote but should be an opening one.

### PR DESCRIPTION
**Pull request by dimaMachina**
fixes https://github.com/silvenon/remark-smartypants/issues/74

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,23 +1,52 @@
...
+    quotes: false,
+  });
+
+  return (tree) => {
+    let allText = "";
+    let startIndex = 0;
+
+    visit(tree, "text", (node) => {
+      allText += node.value;
+    });
```
Comment (2wheeh) [inline index.js]: "inlineCode" doesn't have its text content as "text" children but just has it on value field

```suggestion
    const checkFn = (node) => node.type === "text" || node.type === "inlineCode";
    visit(tree, checkFn, (node) => {
      allText += node.value;
    });
```
same for the second visit.

**Review by 2wheeh — COMMENTED** [review]
and for the test:

```javascript
it("handles quotes around inline codes", async () => {
  const file = await remark()
    .use(smartypants)
    .process(`foo "\`code\`" bar "\`code\` baz"`);
  expect(file.value).toMatchInlineSnapshot(`
    "foo “\`code\`” bar “\`code\` baz”
    "
  `);
});
```

what do you think ?
or should I work this in separated PR after this get merged?

@dimaMachina @silvenon

**Comment by dimaMachina** [comment]
@2wheeh check my last commit

File: index.js (hunk 1 above)
Comment (dimaMachina) [reply index.js]: we can just use `['text', 'inlineCode']`

File: index.js [hunk 2]
```diff
@@ -33,20 +33,24 @@ export default function remarkSmartypants(options) {
   return (tree) => {
     let allText = "";
     let startIndex = 0;
+    const nodes = []
 
-    visit(tree, "text", (node) => {
+    visit(tree, ["text", 'inlineCode'], (node) => {
```
Comment (2wheeh) [inline index.js]:
```suggestion
    visit(tree, ["text", "inlineCode"], (node) => {
```

typo here.

**Review by 2wheeh — COMMENTED** [review]
Thank you for quick iterating !

File: index.js [hunk 3]
```diff
@@ -1,23 +1,56 @@
...
+ * @typedef {import("retext-smartypants").Options} Options
  */
 
 import { retext } from "retext";
 import { visit } from "unist-util-visit";
 import smartypants from "retext-smartypants";
 
 /**
- * remark plugin to implement SmartyPants.
+ * Remark plugin to implement SmartyPants.
```
Comment (silvenon) [inline index.js]: Can you turn this back? I don't want to capitalize it, they keep lowercasing it so I want to as well.
```suggestion
 * remark plugin to implement SmartyPants.
```
Screenshot 2023-12-31 at 11 25 32

File: index.js [hunk 4]
```diff
@@ -1,23 +1,56 @@
 /**
- * @typedef {import('mdast').Root} Root
- * @typedef {import('retext-smartypants').Options} Options
+ * @typedef {import("mdast").Root} Root
+ * @typedef {import("retext-smartypants").Options} Options
```
Comment (silvenon) [inline index.js]: Can you turn off whatever is doing these changes and revert these double quotes in JSDoc comments? I'm not against them, they just don't change anything, so they're needlessly enlarging the diff. Too bad Prettier can't enforce this 🤔

**Review by silvenon — CHANGES_REQUESTED** [review]
Other than the minor changes and fixing new Prettier CI check this looks great, thanks so much for fixing this and adding tests! :heart:

**Comment by silvenon** [comment]
(It's possible that for Prettier you'll need to merge in latest `main`.)

File: index.js [hunk 5]
```diff
@@ -10,14 +10,47 @@ import smartypants from "retext-smartypants";
 /**
  * remark plugin to implement SmartyPants.
  *
- * @type {import('unified').Plugin<[Options?] | void[], Root>}
+ * @type {import("unified").Plugin<[Options?] | void[], Root>}
```
Comment (dimaMachina) [inline index.js]:
```suggestion
 * @type {import('unified').Plugin<[Options?] | void[], Root>}
```

File: index.js [hunk 6]
```diff
@@ -1,6 +1,6 @@
 /**
- * @typedef {import('mdast').Root} Root
- * @typedef {import('retext-smartypants').Options} Options
+ * @typedef {import("mdast").Root} Root
```
Comment (dimaMachina) [inline index.js]:
```suggestion
 * @typedef {import('mdast').Root} Root
```

File: index.js [hunk 7]
```diff
@@ -1,6 +1,6 @@
 /**
  * @typedef {import('mdast').Root} Root
- * @typedef {import('retext-smartypants').Options} Options
+ * @typedef {import("retext-smartypants").Options} Options
```
Comment (dimaMachina) [inline index.js]:
```suggestion
 * @typedef {import('retext-smartypants').Options} Options
```

File: index.js (hunk 4 above)
Comment (dimaMachina) [reply index.js]: Seems it was my local prettier extension, reverted

**Review by silvenon — COMMENTED** [review]
Everything is good, just the Prettier task is failing for the test file, all you need to do is: `npx prettier --write .`

I created a `pre-commit` hook on `main`, if you merged that changed into your branch. It can be installed by running `npm run setup`, that way all your commits will be prettified right upon committing.

**Comment by dimaMachina** [comment]
> Everything is good, just the Prettier task is failing for the test file, all you need to do is: `npx prettier --write .`
> 
> I created a `pre-commit` hook on `main`, if you merged that changed into your branch. It can be installed by running `npm run setup`, that way all your commits will be prettified right upon committing.

finally back to the computer, applied changes, and fixed another possible edge cases with inline code blocks

https://github.com/dimaMachina/remark-smartypants/blob/fe9c3ee4f501f34aae65524866c4e204b70f857d/index.test.js#L53-L71

**Review by silvenon — COMMENTED** [review]
Great! It seems like in the meantime Vitest no longer needs to escape quotes in snapshots (notice that some tests fail), so update that and we're good to go! 🚀

**Comment by dimaMachina** [comment]
> It seems like in the meantime Vitest no longer needs to escape quotes in snapshots (notice that some tests fail), so update that and we're good to go!

locally it was not fail, strange, just removed escaping to see if it fixes tests on CI

**Review by silvenon — APPROVED** [review]

**Comment by silvenon** [comment]
These changes have been published in v2.1.0!
